subsys/net/lib/coap/coap.c

```diff
@@ -1817,23 +1819,44 @@ struct coap_reply *coap_response_received(
 		return NULL;
 	}
 
+	type = coap_header_get_type(response);
 	id = coap_header_get_id(response);
 	tkl = coap_header_get_token(response, token);
 
+	if (type == COAP_TYPE_ACK || type == COAP_TYPE_RESET) {
+		piggybacked = true;
+	}
+
 	for (i = 0, r = replies; i < len; i++, r++) {
 		int age;
 
 		if ((r->id == 0U) && (r->tkl == 0U)) {
 			continue;
 		}
 
-		/* Piggybacked must match id when token is empty */
-		if ((r->id != id) && (tkl == 0U)) {
-			continue;
-		}
+		if (piggybacked) {
+			/* In a piggybacked response, the Message ID of the
+			 * Confirmable request and the Acknowledgment MUST
+			 * match, and the tokens of the response and original
+			 * request MUST match.
+			 */
+			if ((r->id != id || (r->tkl != tkl))) {
+				continue;
+			}
 
-		if (tkl > 0 && memcmp(r->token, token, tkl)) {
-			continue;
+			if (r->tkl > 0) {
+				if (memcmp(r->token, token, r->tkl) != 0) {
+					continue;
+				}
+			}
+		} else {
+			/* In a separate response, just the tokens of the
+			 * response and original request MUST match.
+			 */
+			if ((r->tkl == 0) || (r->tkl != tkl) ||
```

Comment:
If the code-duplicates are intended, then the `(r->tkl == 0)` is wrong and must be removed. That will then also change the results of the tests, as I commended there.

Reply:
Reworked as suggested in earlier comment.